File: tests/suid_sudo_test.rb
#!/usr/bin/ruby -T
if __FILE__ != $0
  p "run directly."
  exit(1)
end
require File.absolute_path("../suid_sudo", File.dirname(__FILE__)).untaint
# require_relative claims insecure operation

include SUID_SUDO::INCLUDE
p suid_emulate(sudo_wrap:true, use_shebang:false, pass_env:['TESTVAR'])

cmd = ARGV[0] || "0"

def _ids
  p [[Process::Sys::getuid, Process::Sys::geteuid],
     [Process::Sys::getgid, Process::Sys::getegid],
     Process.groups]
end
def _print_ids
  p _ids
end

case cmd
when '0'
  _print_ids
  p ARGV
  ENV.each {|k,v| print "#{k}=#{v}\n" }
when '1'
  drop_privileges_forever
  _print_ids
  p ARGV
  ENV.each {|k,v| print "#{k}=#{v}\n" }
when 'setuid'
  print ("be USER ")
  temporarily_as_user()
  _print_ids

  print ("be ROOT")
  temporarily_as_root()
  _print_ids

  print ("be REAL_ROOT")
  temporarily_as_real_root()
  _print_ids

  print ("be USER")
  temporarily_as_user()
  _print_ids

  print ("drop to USER")
  drop_privileges_forever()
  _print_ids

  begin
    print ("be ROOT")
    temporarily_as_root()
    _print_ids
    print ("SHOULD FAILED!\n")
  rescue SUIDPrivilegesSettingError => e
    print("...good to be failed: #{e}\n")
  end
when 'temp'
  p "temporary_in_user"
  temporarily_as_user {
    _print_ids
  }
  p "returned"
  _print_ids
when 'temp_root'
  p "temporary_in_root"
  temporarily_as_root {
    _print_ids
  }
  p "returned"
  _print_ids
when 'system'
  p "system"
  p spawn_in_privilege(:system, :drop_privileges_forever, "/usr/bin/id")
when 'system_ba'
  p "system"
  p spawn_in_privilege(:system, :drop_privileges_foreverr, "/usr/bin/id")
when 'system_proc'
  p "system"
  p spawn_in_privilege(:system, method(:drop_privileges_forever), "/usr/bin/id")
when 'system_e'
  p "system"
  p spawn_in_privilege(:system, :drop_privileges_forever, "/dev/null")
when 'system_e2'
  p "system"
  p spawn_in_privilege(:system, :drop_privileges_forever, "/bin/false", exception:true)
when 'system_e2c'
  p system("/bin/false", exception:true)
when 'spawn'
  p "spawn"
  r = spawn_in_privilege(:spawn, :drop_privileges_forever, "/usr/bin/id")
  p r
  p Process.waitpid2(r)
when 'spawn_e'
  p "spawn"
  p spawn_in_privilege(:spawn, :drop_privileges_forever, "/dev/null")
when 'subproc'
  p "call_in_subprocess"
  result = run_in_subprocess {
    drop_privileges_forever
    _print_ids
    _ids
  }
  p ["result", result]
  _print_ids
when 'subproc_fd'
  p "call_in_subprocess"
  begin
    result = run_in_subprocess {
      exec("ls", "-lL", "/proc/self/fd")
      raise "BAD"
    }
    p result
  rescue SUIDSubprocessExecutionError => e
    p "OK #{e}"
  end
when 'subproc_fd_e1'
  # ensure above raise "BAD" is not captured
  p "call_in_subprocess"
  begin
    result = run_in_subprocess {
      raise "BAD"
    }
    p result
  rescue SUIDSubprocessExecutionError => e
    exit "BAD #{e.inspect}"
  rescue RuntimeError => e
    p "OK #{e.inspect}"
  end
when 'subproc_cs'
  p "call_in_subprocess"
  result = run_in_subprocess {
    [1, {2 => 3, 4 => 5}, "abc \u3001\u3002 cde"]
  }
  p result
when 'subproc_error1'
  p "call_in_subprocess"
  begin
    result = run_in_subprocess {
      ([1,2]).fetch(2)
      return "BAD"
    }
    p result
  rescue IndexError => e
    p "OK #{e}"
  end
when 'subproc_error1e'
  p "call_in_subprocess"
  begin
    result = run_in_subprocess {
      raise SUIDHandlingError::new("test") # non-wrappable exception
    }
    p result
  rescue WrappedSubprocessError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error1e2'
  p "call_in_subprocess"
  begin
    result = run_in_subprocess {
      Math::sqrt(-4)
    }
    p result
  rescue Math::DomainError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error2'
  begin
    result = run_in_subprocess {
      exit(255)
    }
  rescue SUIDSubprocessExecutionError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error2_1'
  begin
    result = run_in_subprocess {
      exit!(255)
    }
  rescue SUIDSubprocessExecutionError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error3'
  begin
    result = run_in_subprocess {
      drop_privileges_forever()
      exec("/usr/bin/id")
    }
  rescue SUIDSubprocessExecutionError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error4'
  begin
    result = run_in_subprocess {
      exec("/dev/null")
    }
  rescue SystemCallError => e
    p "OK #{e.inspect}"
  end
when 'subproc_error5'
  begin
    result = run_in_subprocess {
      drop_privileges_forever()
      break
    }
  rescue SUIDSubprocessExecutionError => e
    p "OK #{e.inspect}"
  end
else
  p "unknown command"
end

File: tests/suid_sudo_test.py
#!/usr/bin/python3 -I
if __name__ != '__main__': raise RuntimeError("Should not be imported.")

import sys
import os.path
sys.path.insert(0, os.path.dirname(sys.argv[0]) + "/..")
import suid_sudo
from suid_sudo import *
_SuidStatus = suid_sudo._SuidStatus

import os
def print_ids():
    print (os.getresuid(), os.getresgid(), os.getgroups())

suid_emulate(sudo_wrap=True, inherit_flags=True, realroot_ok=True, nonsudo_ok=True,
             pass_env = [], showcmd_opts=True)

cmd = sys.argv[1] if len(sys.argv) >= 2 else "0"

if cmd == "0":
    print (sys.version)
    print_ids()
    print (sys.argv)
    print (_SuidStatus._status)
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

elif cmd == "1":
    drop_privileges_forever()
    print (sys.version)
    print_ids()
    print (sys.argv)
    print (_SuidStatus._status)
    for k in os.environ:
        print ("%s=%s" % (k, os.environ[k]))

elif cmd == "setuid":
    print ("be USER ")
    temporarily_as_user()
    print_ids()

    print ("be ROOT")
    temporarily_as_root()
    print_ids()

    print ("be REAL_ROOT")
    temporarily_as_real_root()
    print_ids()

    print ("be USER")
    temporarily_as_user()
    print_ids()

    print ("drop to USER")
    drop_privileges_forever()
    print_ids()

    try:
        print ("be ROOT")
        temporarily_as_root()
        print_ids()
        print ("SHOULD FAILED!")
        raise
    except (SUIDPrivilegesSettingError, OSError) as e:
        print("...good to be failed: %r" % e)

elif cmd == "temp":
    print("temporary_in_user")
    with temporarily_as_user():
        print_ids()

    print ("returned")
    print_ids()

elif cmd == "subcall":
    print ("subprocess as real user")
    import subprocess
    subprocess.call(args=["/usr/bin/id"],
                    preexec_fn=drop_privileges_forever)

    print ("returned")
    print_ids()

elif cmd == "subproc":
    print ("call_in_subprocess")
    def _():
        drop_privileges_forever()
        return (os.getresuid(), os.getresgid(), os.getgroups())
    print(call_in_subprocess(_))

    print ("parent")
    print_ids()

    print ("call_in_subprocess_2")
    @call_in_subprocess
    def rval():
        drop_privileges_forever()
        return (os.getresuid(), os.getresgid(), os.getgroups())
    print(rval)

    print ("parent")
    print_ids()

elif cmd == "subproc_fd":
    print ("call_in_subprocess")
    def _():
        os.execlp("ls", "ls", "-lL", "/proc/self/fd")
        # you should not see a pipe in this directory entry
    try:
        print(call_in_subprocess(_))
    except SUIDSubprocessError as e:
        print("OK: exception %r" % (e,))

elif cmd == "subproc_error1":
    print ("call_in_subprocess")
    def _():
        drop_privileges_forever()
        ([1,2])[2]
        return ("BAD", os.getresuid(), os.getresgid(), os.getgroups())
    try:
        print(call_in_subprocess(_))
    except IndexError as e:
        print("OK: exception %r" % (e,))

elif cmd == "subproc_error2":
    print ("call_in_subprocess")
    def _():
        os._exit(255)
    try:
        print(call_in_subprocess(_))
    except SUIDSubprocessExecutionError as e:
        print("OK: exception %r" % (e,))

elif cmd == "subproc_error2_1":
    print ("call_in_subprocess")
    def _():
        exit(100)
    try:
        print(call_in_subprocess(_))
    except WrappedSubprocessError as e:
        # SystemExit is wrapped (otherwise, the parent terminates!)
        print("OK: exception %r" % (e,))

elif cmd == "subproc_error3":
    print ("call_in_subprocess")
    def _():
        os.execl("/usr/bin/id", "/usr/bin/id")
    try:
        print(call_in_subprocess(_))
    except SUIDSubprocessExecutionError as e:
        print("OK: exception %r" % (e,))

elif cmd == "subproc_error4":
    print ("call_in_subprocess")
    def _():
        os.execl("/dev/null", "/dev/null")
    try:
        print(call_in_subprocess(_))
    except OSError as e:
        print("OK: exception %r" % (e,))

else:
    print("unknown test command")
